drivers/media/platform/bcm2835/bcm2835-unicam.c

```diff
@@ -0,0 +1,2825 @@
+	    !v4l2_subdev_has_op(unicam->sensor, pad, get_selection))
+		v4l2_disable_ioctl(&node->video_dev, VIDIOC_G_SELECTION);
+
+	ret = video_register_device(vdev, VFL_TYPE_VIDEO, -1);
+	if (ret) {
+		unicam_err(unicam, "Unable to register video device %s\n",
+			   vdev->name);
+		return ret;
+	}
+
+	/*
+	 * Acquire a reference to unicam, which will be released when the video
+	 * device will be unregistered and userspace will have closed all open
+	 * file handles.
+	 */
+	unicam_get(unicam);
+	node->registered = true;
+
+	if (pad_id != METADATA_PAD || unicam->sensor_embedded_data) {
+		ret = media_create_pad_link(&unicam->sensor->entity, pad_id,
+					    &node->video_dev.entity, 0,
+					    MEDIA_LNK_FL_ENABLED |
+					    MEDIA_LNK_FL_IMMUTABLE);
+		if (ret)
+			unicam_err(unicam, "Unable to create pad link for %s\n",
+				   vdev->name);
+	}
+
+	return ret;
+}
+
+static void unregister_nodes(struct unicam_device *unicam)
+{
+	unsigned int i;
+
+	for (i = 0; i < ARRAY_SIZE(unicam->node); i++) {
+		struct unicam_node *node = &unicam->node[i];
+
+		if (node->dummy_buf_cpu_addr) {
+			dma_free_coherent(&unicam->pdev->dev, DUMMY_BUF_SIZE,
+					  node->dummy_buf_cpu_addr,
+					  node->dummy_buf_dma_addr);
+		}
+
+		if (node->registered) {
+			node->registered = false;
+			video_unregister_device(&node->video_dev);
+		}
+	}
+}
+
+static int unicam_probe_complete(struct unicam_device *unicam)
+{
+	int ret;
+
+	unicam->v4l2_dev.notify = unicam_notify;
+
+	unicam->sensor_config = v4l2_subdev_alloc_pad_config(unicam->sensor);
+	if (!unicam->sensor_config)
+		return -ENOMEM;
+
+	unicam->sensor_embedded_data = (unicam->sensor->entity.num_pads >= 2);
+
+	ret = register_node(unicam, &unicam->node[IMAGE_PAD],
+			    V4L2_BUF_TYPE_VIDEO_CAPTURE, IMAGE_PAD);
+	if (ret) {
+		unicam_err(unicam, "Unable to register image video device.\n");
+		goto unregister;
+	}
+
+	ret = register_node(unicam, &unicam->node[METADATA_PAD],
+			    V4L2_BUF_TYPE_META_CAPTURE, METADATA_PAD);
+	if (ret) {
+		unicam_err(unicam, "Unable to register metadata video device.\n");
+		goto unregister;
+	}
+
+	ret = v4l2_device_register_ro_subdev_nodes(&unicam->v4l2_dev);
+	if (ret) {
+		unicam_err(unicam, "Unable to register subdev nodes.\n");
+		goto unregister;
+	}
+
+	/*
+	 * Release the initial reference, all references are now owned by the
+	 * video devices.
+	 */
+	unicam_put(unicam);
+	return 0;
+
+unregister:
+	unregister_nodes(unicam);
+	unicam_put(unicam);
+
+	return ret;
+}
+
+static int unicam_async_complete(struct v4l2_async_notifier *notifier)
+{
+	struct unicam_device *unicam = to_unicam_device(notifier->v4l2_dev);
+
+	return unicam_probe_complete(unicam);
+}
+
+static const struct v4l2_async_notifier_operations unicam_async_ops = {
+	.bound = unicam_async_bound,
+	.complete = unicam_async_complete,
+};
+
+static int of_unicam_connect_subdevs(struct unicam_device *dev)
+{
+	struct platform_device *pdev = dev->pdev;
+	struct v4l2_fwnode_endpoint ep = { 0 };
```

Comment:
I get a build warning on this line:
```
drivers/media/platform/bcm2835/bcm2835-unicam.c: In function 'of_unicam_connect_subdevs':
drivers/media/platform/bcm2835/bcm2835-unicam.c:2582:9: warning: missing braces around initializer [-Wmissing-braces]
  struct v4l2_fwnode_endpoint ep = { 0 };
         ^
drivers/media/platform/bcm2835/bcm2835-unicam.c:2582:9: warning: (near initialization for 'ep.base') [-Wmissing-braces]
```

Reply:
It's happy with ` = { { 0 } };`.

Reply:
Isn't `= { }` the correct initialiser?
Including the `0` means the first element of structure needs to be assignable to zero which isn't the case with many types (including sub-structures)

Reply:
I think the problem is that the first element of a v4l2_fwnode_endpoint is another struct -  fwnode_endpoint, which has a scalar value as the first element.

Reply:
Yes but doesn't `= {}` avoid the warning?
It's a gcc construct, but is widely used in the kernel.

Reply:
Are you saying you don't get a warning?

Reply:
I don't with `= {}` I do with `={0}`

Reply:
```
phil@buildbot:~/pi/linux$ git grep '= { { 0 } };' |  wc
     54     533    5135
phil@buildbot:~/pi/linux$ git grep '= { { { 0 } } };' |  wc
      4      49     396
```

Reply:
```
$ git grep '= { };' | wc
    573    4043   41864
```

Reply:
= {0} is the correct syntax to zero an entire structure to zero, but there is/was a known bug in gcc that meant it failed to compile in some cases. I thought it had been fixed though. I was on the distribution list for the fix because I reported it at some point in the ancient past.

Reply:
```
$ git grep '= {};' | wc
   1754   10450  132323
```

Reply:
```
phil@buildbot:~/pi/linux$ git grep '= {};' |  wc
   1754   10450  132323
```

Reply:
You win. It's not an extension I'd noticed before, but it makes more sense than the alternative.

Reply:
It comes from the version of the patch that was being upstreamed, and they tend to use a later gcc than we appear to.
{0} is meant to zero the whole struct.

Reply:
This isn't going to make for a particularly tricky merge conflict if it does change upstream.

Reply:
I'm happy to create an additional downstream patch here to use memset instead to avoid the build warning, but I'm hoping Laurent will otherwise pick up the patches to create the next patchset to send to the list, and they'll want ```= { 0 }```